test/e2e/shared/aws_helpers.go

```diff
@@ -0,0 +1,176 @@
+	"github.com/aws/aws-sdk-go/aws/client"
+	rgapi "github.com/aws/aws-sdk-go/service/resourcegroupstaggingapi"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	infrav1 "sigs.k8s.io/cluster-api-provider-aws/api/v1beta1"
+)
+
+type LoadBalancerType string
+
+var (
+	LoadBalancerTypeELB = LoadBalancerType("elb")
+	LoadBalancerTypeALB = LoadBalancerType("alb")
+	LoadBalancerTypeNLB = LoadBalancerType("nlb")
+)
+
+type WaitForLoadBalancerToExistForServiceInput struct {
+	AWSSession       client.ConfigProvider
+	ServiceName      string
+	ServiceNamespace string
+	ClusterName      string
+	Type             LoadBalancerType
+}
+
+func WaitForLoadBalancerToExistForService(ctx context.Context, input WaitForLoadBalancerToExistForServiceInput, intervals ...interface{}) {
+	By(fmt.Sprintf("Waiting for AWS load balancer of type %s to exist for service %s/%s", input.Type, input.ServiceNamespace, input.ServiceName))
+
+	Eventually(func() bool {
+		arns, err := GetLoadBalancerARNs(ctx, GetLoadBalancerARNsInput{ //nolint: gosimple
+			AWSSession:       input.AWSSession,
+			ServiceName:      input.ServiceName,
+			ServiceNamespace: input.ServiceNamespace,
+			ClusterName:      input.ClusterName,
+			Type:             input.Type,
+		})
+		if err != nil {
+			fmt.Fprintf(GinkgoWriter, "error getting loadbalancer arns: %v\n", err)
+
+			return false
+		}
+		if len(arns) == 0 {
+			return false
+		}
+
+		return true
+	}, intervals...).Should(BeTrue(), "failed to wait for loadbalancer")
+}
+
+type GetLoadBalancerARNsInput struct {
+	AWSSession       client.ConfigProvider
+	ServiceName      string
+	ServiceNamespace string
+	ClusterName      string
+	Type             LoadBalancerType
+}
+
+func GetLoadBalancerARNs(ctx context.Context, input GetLoadBalancerARNsInput) ([]string, error) {
+	By(fmt.Sprintf("Getting AWS load balancer ARNs of type %s for service %s/%s", input.Type, input.ServiceNamespace, input.ServiceName))
+
+	serviceTag := infrav1.ClusterAWSCloudProviderTagKey(input.ClusterName)
+	tags := map[string][]string{
+		"kubernetes.io/service-name": {fmt.Sprintf("%s/%s", input.ServiceNamespace, input.ServiceName)},
+		serviceTag:                   {string(infrav1.ResourceLifecycleOwned)},
+	}
+	descInput := &DescribeResourcesByTagsInput{
+		AWSSession: input.AWSSession,
+		Tags:       tags,
+	}
+
+	descOutput, err := DescribeResourcesByTags(*descInput)
+	if err != nil {
+		fmt.Fprintf(GinkgoWriter, "error querying resources by tags: %v\n", err)
+		return nil, fmt.Errorf("describing resource tags: %w", err)
+	}
+
+	matchingARNs := []string{}
+	for i := range descOutput.ARNs {
+		resARN := descOutput.ARNs[i]
+		parsedArn, err := arn.Parse(resARN)
+		if err != nil {
+			fmt.Fprintf(GinkgoWriter, "error parsing arn %s: %v\n", resARN, err)
+			return nil, fmt.Errorf("parsing resource arn %s: %w", resARN, err)
+		}
+
+		if parsedArn.Service != "elasticloadbalancing" {
+			continue
+		}
+
+		switch input.Type {
+		case LoadBalancerTypeALB:
```

Comment:
This is a bit weird. If you already have the type, is it necessary in the test to check the arn as well?
Wouldn't you make sure in the test that once you say it's an NLB it IS actually an NLB? :)

Or am I misunderstanding what this does?

Reply:
This function is used to get all the loadbalancer ARNs for a specific type of load balancer....which is supplied.  Using the resource tagging api like this means i only need 1 function.